locallib.php

```diff
@@ -965,72 +966,99 @@ public function distrubute_choices() {
+            $data = new stdClass();
+            $data->groupingid = $groupingid;
+            $data->ratingallocateid = $this->ratingallocateid;
+            $this->db->insert_record(this_db\ratingallocate_id_grouping::TABLE, $data);
+
         } else {
-            $groupingid = $grouping->id;
+            // If there is already a grouping for this allocation assign the corresponing id to groupingid.
+            $groupingids = $this->db->get_records_select(this_db\ratingallocate_id_grouping::TABLE,
+                            "ratingallocateid=$this->ratingallocateid",
+                            null,
+                            '',
+                            'id, groupingid');
+            // There should only be one entry in groupingids.
+            $groupingid = 0;
+            foreach ($groupingids as $groupingidsentry) {
+                $groupingid = $groupingidsentry->groupingid;
+            }
         }
 
-        $groupidentifierfromchoiceid = function ($choiceid) {
-            return ratingallocate_MOD_NAME . '_c_' . $choiceid;
-        };
-
         $choices = $this->get_choices_with_allocationcount();
 
-        // Make a new array containing only the identifiers of the choices.
-        $choiceids = array();
-        foreach ($choices as $id => $choice) {
-            $choiceids[$groupidentifierfromchoiceid($choice->id)] = array('key' => $id);
-        }
+        // Loop through existing choices.
+        foreach ($choices as $choice) {
+            if ($this->db->record_exists(this_db\ratingallocate_choices::TABLE,
+                    ['id' => $choice->id])) {
 
-        // Dind all associated groups in this grouping.
-        $groups = groups_get_all_groups($this->course->id, 0, $groupingid);
+                // Checks if there is already a group for this choice.
+                if ($this->db->record_exists(this_db\ratingallocate_choice_group::TABLE,
+                    ['choiceid' => $choice->id])) {
 
-        // Loop through the groups in the grouping: if the choice does not exist anymore -> delete.
-        // Otherwise mark it.
-        foreach ($groups as $group) {
-            if (array_key_exists($group->idnumber, $choiceids)) {
-                // Group exists, mark.
-                $choiceids[$group->idnumber]['exists'] = true;
-                $choiceids[$group->idnumber]['groupid'] = $group->id;
-            } else {
-                // Delete group $group->id.
-                groups_delete_group($group->id);
-            }
-        }
+                    // Get the group from the choice_group Table.
+                    $groupids = $this->db->get_records_select(this_db\ratingallocate_choice_group::TABLE,
+                        "choiceid=$choice->id",
+                        null,
+                        '',
+                        'id, groupid');
+                    // Only one object in groupids because there should only be one entry in the table with this choiceid.
```

Comment:
If there should be only one object available, is there corresponding error handling for that? Then get_record() should be enough here and there is no need for a loop. But there should be sufficient checks when adding to the database to guarantee there are no duplicate entries for the same choice.